File: geeta book app/lib/shared/domain/models/user_progress.dart
import 'package:freezed_annotation/freezed_annotation.dart';

part 'user_progress.freezed.dart';
part 'user_progress.g.dart';

@freezed
class UserProgress with _$UserProgress {
  const factory UserProgress({
    required String userId,
    required int totalChaptersCompleted,
    required int totalVersesCompleted,
    required int totalActivitiesCompleted,
    @Default({}) Map<String, ChapterProgress> chapterProgress,
    DateTime? lastUpdated,
  }) = _UserProgress;

  factory UserProgress.fromJson(Map<String, dynamic> json) =>
      _$UserProgressFromJson(json);
}

@freezed
class ChapterProgress with _$ChapterProgress {
  const factory ChapterProgress({
    required String chapterId,
    required int versesCompleted,
    required int totalVerses,
    required int activitiesCompleted,
    required int totalActivities,
    @Default(false) bool isCompleted,
    DateTime? completedAt,
  }) = _ChapterProgress;

  factory ChapterProgress.fromJson(Map<String, dynamic> json) =>
      _$ChapterProgressFromJson(json);
}

File: admin-server/src/models/UserProgress.js
const mongoose = require('mongoose');

const userProgressSchema = new mongoose.Schema({
  userId: {
    type: String,
    required: true,
    unique: true,
    index: true,
  },
  totalChaptersCompleted: {
    type: Number,
    default: 0,
  },
  totalVersesCompleted: {
    type: Number,
    default: 0,
  },
  totalActivitiesCompleted: {
    type: Number,
    default: 0,
  },
  chapterProgress: {
    type: Map,
    of: {
      versesCompleted: { type: Number, default: 0 },
      totalVerses: { type: Number, required: true },
      activitiesCompleted: { type: Number, default: 0 },
      totalActivities: { type: Number, default: 6 },
      isCompleted: { type: Boolean, default: false },
      completedAt: { type: Date },
    },
    default: {},
  },
  completedVerses: [{
    verseId: String,
    completedAt: Date,
  }],
  completedActivities: [{
    activityId: String,
    completedAt: Date,
  }],
}, {
  timestamps: true,
});

module.exports = mongoose.model('UserProgress', userProgressSchema);
